Bluetooth: Controller: fix for ISO-AL EBQ Tests and Support TX Sync

niag-Oticon opened the pull request:
Bluetooth: Controller: fix for ISO-AL EBQ Tests and Support TX Sync

Changes:
-- IAL/CIS/FRA/PER/BI-01-C & IAL/CIS/FRA/PER/BI-02-C verified data length in segment header
-- Implemented TX Sync command
-- IALCISUNFPERBV(12|36)C & IALCISFRAPERBV(13|38|49)C buffering on SDU fragments to get correct SDU size in HCI ISO data packet, load field header
-- IALCISFRAPERBV(07|31|47)C packing multiple SDU fragments into one PDU
-- Disabled ISO-AL unit tests (tests to be aligned and submitted as a separate PR)

niag-Oticon committed:
Bluetooth: controller: Included guard for lost data in framed PDUs

Improvement for IAL/CIS/FRA/PER/BI-01-C & IAL/CIS/FRA/PER/BI-02-C
-- Test appears to fail as the transmitted PDU is packed with less data
   than is indicated in the segment headers.
-- Guards were included to verify that sufficient data exists in the PDU
   before extracting the segment.

General
-- Updated timestamps in error SDUs to better approximate the expected
   timestamps if time information is missing.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
tests: Bluetooth: controller: Disabled ISO-AL Unit Tests

Tests will be restructured in a future commit. Temporarily disabled to
partition the changes to allow tests to be a separate PR.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_iso.c:
```diff
@@ -1258,6 +1290,31 @@ void ull_iso_lll_ack_enqueue(uint16_t handle, struct node_tx_iso *node_tx)
+
+void ull_iso_lll_event_prepare(uint16_t handle, uint64_t event_count)
+{
+	struct ll_iso_datapath *dp = NULL;
+
+	if (IS_CIS_HANDLE(handle)) {
+		struct ll_conn_iso_stream *cis;
```
```suggestion
	if (IS_CIS_HANDLE(handle)) {
		const struct ll_iso_datapath *dp = NULL;
		const struct ll_conn_iso_stream *cis;
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_iso/src/main.c:
```diff
@@ -106,12 +106,13 @@ static isoal_status_t test_sink_sdu_alloc(const struct isoal_sink    *sink_ctx,
-	printk("Vendor sink SDU len %u, seq_num %u, ts %u\n",
-		sink_ctx->sdu_production.sdu_written, valid_sdu->seqn,
-		valid_sdu->timestamp);
+	printk("Vendor sink SDU fragment len %u / %u, seq_num %u, ts %u\n",
+		sdu_frag->sdu_frag_size, sdu->total_sdu_size, sdu_frag->sdu.seqn,
```
`printk` says "len", but variable is named "size". While these are very similar, I suggest to use same terminology

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -28,6 +28,23 @@
 #include "common/log.h"
 #include "hal/debug.h"
 
+/* Minimum number of bytes that would make inserting a new segment into a PDU
+ * worthwhile.
+ */
+#if defined(CONFIG_BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN)
+#define ISOAL_SEGMENT_MIN_PAYLOAD         CONFIG_BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN
+#else
+#define ISOAL_SEGMENT_MIN_PAYLOAD         (1)
+#endif /* CONFIG_BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN */
```
I'm curious:
Why is a TX Kconfig used here? 
And in what case is `CONFIG_BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN` not defined where we have ISO-AL?

niag-Oticon replied:
I think your observation is mostly correct. If the ISO-AL supports TX then it should be that `CONFIG_BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN` is defined with some value. So it should be possible to rely on the default.

Thalley replied:
But is `ISOAL_SEGMENT_MIN_PAYLOAD` only used for TX?

niag-Oticon replied:
Yes it is only used for TX. That is the threshold that would make inserting a new segment worthwhile when packing PDUs.

Thalley replied:
OK. The name of the macro `ISOAL_SEGMENT_MIN_PAYLOAD` doesn't really explicitely state this, hence my confusion. 

If `ISOAL_SEGMENT_MIN_PAYLOAD` is for TX only, this is fine, although the name of the macro may need to more explicit perhaps. 

But if `ISOAL_SEGMENT_MIN_PAYLOAD` is for TX only, then why do we need this macro at all? If we can TX in the ISO-AL, wouldn't `CONFIG_BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN` always be defined?

niag-Oticon replied:
Sorry about the omission. I have updated it now. Yes it's not strictly necessary. It was a remnant of how it evolved.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -110,6 +126,7 @@ static isoal_status_t isoal_sink_allocate(isoal_sink_handle_t *hdl)
 static void isoal_sink_deallocate(isoal_sink_handle_t hdl)
 {
 	isoal_global.sink_allocated[hdl] = ISOAL_ALLOC_STATE_FREE;
+	memset(&isoal_global.sink_state[hdl], 0, sizeof(struct isoal_sink));
```
```suggestion
	(void)memset(&isoal_global.sink_state[hdl], 0, sizeof(struct isoal_sink));
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.h:
```diff
@@ -8,6 +8,11 @@
 #include <zephyr/toolchain.h>
 
+#if defined(CONFIG_BT_CTLR_ISO_RX_SDU_BUFFERS) && (CONFIG_BT_CTLR_ISO_RX_SDU_BUFFERS > 0)
+#define ISOAL_CONFIG_BUFFER_RX_SDUS_ENABLE
```
`XXX_CONFIG_YYY` seems like a poor name, as it can be mistaken for a Kconfig with a typo.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -327,6 +345,127 @@ static isoal_status_t isoal_rx_allocate_sdu(struct isoal_sink *sink,
+			/* Collated information is not reliable if there is an
+			 * error in the sequence of the fragments.
+			 */
+			for (int i = 0; i < next_write_indx; i++) {
```
```suggestion
			for (uint8_t i = 0; i < next_write_indx; i++) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -327,6 +345,127 @@ static isoal_status_t isoal_rx_allocate_sdu(struct isoal_sink *sink,
+			}
+		}
+
+		for (int i = 0; i < next_write_indx; i++) {
```
```suggestion
		for (uint8_t i = 0; i < next_write_indx; i++) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -942,7 +1123,40 @@ static isoal_status_t isoal_source_allocate(isoal_source_handle_t *hdl)
+	}
+
 	isoal_global.source_allocated[hdl] = ISOAL_ALLOC_STATE_FREE;
+	memset(&isoal_global.source_state[hdl], 0, sizeof(struct isoal_source));
```
```suggestion
	(void)memset(&isoal_global.source_state[hdl], 0, sizeof(struct isoal_source));
```

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1276,6 +1494,31 @@ static isoal_status_t isoal_tx_unframed_produce(struct isoal_source *source,
+		 * this seems to be the best candidate.
+		 */
+		if (actual_event > tx_sdu->target_event) {
+			actual_grp_ref_point = tx_sdu->grp_ref_point +
+				((actual_event - tx_sdu->target_event) * session->iso_interval *
+					ISO_INT_UNIT_US);
```
Any risk that they may overflow? Looks like we are adding a uint64_t times `session->iso_interval * ISO_INT_UNIT_US)` to a `uint32_t` without any checks of whether the result can be stored.

niag-Oticon replied:
First off the event count is only 39 bits of the 64. 64 bits were used for convenience.

Even so, yes this will overflow but that overflow is expected. The reason for that is that the `grp_ref_point` here is derived from the controller's clock and that is based off a 32-bit value. In Oticon's case a tick is 1/16 of a microsecond so we effectively have only 28 bits for the microsecond range. That is also going to overflow and wrap around. Similarly the time-stamp field according to the specs is also only 32-bits. The important thing here is that the derived reference point and any calculations based on those are accurate across the wrap around and comparable to the controller's clock in the microsecond range.

Thalley replied:
Alright. If an overflow is possible by design, I would suggest to add a small comment about that, as that can be a hard issue to debug otherwise :)

niag-Oticon replied:
I did find a problem in the time-offset calculation in the framed case that would have been affected if the reference point was to wrap around, so I changed the intermediate reference point to 64-bits. Thanks for pointing this out.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1700,6 +1941,41 @@ static isoal_status_t isoal_tx_framed_produce(struct isoal_source *source,
+	source = &isoal_global.source_state[source_hdl];
+	session = &source->session;
+	pp = &source->pdu_production;
+	last_event_payload = (session->burst_number * (event_count + 1)) - 1;
```
Any risk over overflow here?

niag-Oticon replied:
This should not overflow 8-bits BN x 39-bits event count. Once again the 64-bits is for convenience.

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1749,4 +2042,72 @@ void isoal_tx_pdu_release(isoal_source_handle_t source_hdl,
+				      uint32_t *timestamp,
+				      uint32_t *offset)
+{
+	struct isoal_source_session *session;
+
+	if (isoal_check_source_hdl_valid(source_hdl) == ISOAL_STATUS_OK) {
+
```
```suggestion

	if (isoal_check_source_hdl_valid(source_hdl) == ISOAL_STATUS_OK) {
		const struct isoal_source_session *session;

```

niag-Oticon committed:
tests: bluetooth: controller: updated BSIM ISO ISO-AL emit callback I/F

Updated callback function parameters to match new interface parameters.

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1276,6 +1492,31 @@ static isoal_status_t isoal_tx_unframed_produce(struct isoal_source *source,
+		/* Get actual event for this payload number */
+		actual_event = pp->payload_number / session->burst_number;
+
+		/* Get group reference point for this PDU based on the actual
+		 * event being set. This might introduce some errors as the
+		 * group refernce point for future events could drift. However
+		 * as the time offset calculation requires an absolute value,
+		 * this seems to be the best candidate.
+		 */
```
Agreed, the changes that I made have caused the scope of this PR to creep, so I would like to propose that I will return it to the original scope with the comment we discussed.

There is a potential omission in considering the possibility of a wrap around in the framed TX, but this is in the code that is already upstream. So I will create a separate issue for that so that it can be  dealt with it in a separate PR.

Would that be acceptable?

Thalley replied:
I'd like get input from other reviewers on this. I have a preference to fix this in this issue, as introducing new bugs in a `fix` commit just doesn't seem right :D

niag-Oticon replied:
I'm not certain that we are exactly on the same page here. :-) But yes lets gather the feedback from the other reviewers as well and then see how I can best address it.

mtpr-ot replied:
> I'd like get input from other reviewers on this. I have a preference to fix this in this issue, as introducing new bugs in a `fix` commit just doesn't seem right :D

I don't see this as "introducing new bugs". The comment in the code is perhaps unclear, but what is actually going on is that the code predicts the best possible group reference point by multiplying with the ISO interval, knowing that this is going to be slightly off due to drift. This is however by design, and the reference point is "corrected" along the way.

niag-Oticon committed:
Bluetooth: Controller: implemented HCI_LE_Read_ISO_TX_Sync in ISO-AL

Implemented:
-- Storing information required to service HCI_LE_Read_ISO_TX_Sync
   request
-- Implemented interface to retrieve information
-- Cleared ISO-AL source and sink state on deallocation
-- Renamed cig_ref_point to grp_ref_point as it serves a common purpose
   in both CIS and BIS

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
Bluetooth: Controller: Implemented buffering of SDU fragments in ISO-AL

Implements:
-- SDU fragments can be buffered in the ISO-AL so that the correct size
   of the SDU can be computed such that it can be released for inclusion
   in the first fragment's header
-- Changed SDU emit interface to allow the same structure to be used to
   buffer as well as release fragments
-- New configuration BT_CTLR_ISO_RX_SDU_BUFFERS to specify the number of
   SDU buffers for each sink
   (Buffering disabled when BT_CTLR_ISO_RX_SDU_BUFFERS=0)

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

niag-Oticon committed:
Bluetooth: controller: Implemented PDU release timeout

Implemented:
-- Combining segments from multiple SDUs in a single PDU for TX
-- Release of PDU based on remaining capacity below threshold
-- Release of any PDUs in production on receiving a timeout for the
   upcoming event
-- Release of PDUs in production if source if destroyed
-- Synchronization flag to exclude multiple contexts in ISO-AL
-- New Kconfig BT_CTLR_ISO_TX_SEG_PLAYLOAD_MIN to allow definition of
   minimum TX segment size

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

mtpr-ot reviewed (approved):
Looks good, and passes EBQ tests.

cvinayak reviewed (approved):
QUICK-ACK, Limited knowledge of ISOAL, not strong change requests  from my side.
